CLI: Include an editor

roosterfish opened the pull request:
Fixes https://github.com/canonical/microcloud/issues/500.

Add a minimal `vim.tiny` to allow commands like `microcloud cluster recover` to spawn an interactive editor for editing the temporary file.
This increases the overall size of the snap to around 11MB (from 10MB).

roosterfish committed:
snapcraft: Install vim.tiny and keep defaults.vim

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish committed:
snapcraft/commands/microcloud: Set EDITOR env variable to vim.tiny

Also configure the VIM root to prevent an error message caused by trying
to load a default config file not included in 'vim-tiny'.

In addition restrict vim.tiny to only allow writing to the specified file
and don't load extra configuration files.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

simondeziel commented on snapcraft.yaml:
```diff
@@ -84,10 +84,26 @@ parts:
+      - bin/nano
+      - etc/nanorc
+      # Only prime the configuration for yaml.
+      - usr/share/nano/yaml.nanorc
```
I like that tweak, thanks!

simondeziel commented on snapcraft/commands/microcloud:
```diff
@@ -7,4 +7,7 @@ if [ -x "${SNAP_COMMON}/microcloud.debug" ]; then
 fi
 
+# Use nano editor for interactive edits.
+export EDITOR="nano -f /snap/microcloud/current/etc/nanorc"
```
Would you mind throwing `--restricted` in there as well? See https://man.archlinux.org/man/nano.1 for what it does.
This was used by LXD before we completely removed the bundled `nano`.

roosterfish replied:
Done, thx for the suggestion, wasn't aware of the flag!

simondeziel reviewed (commented):
I asked the snap folks if there was a better way than bundling one editor and apparently there is no better way :/

This means we'll have to republish `microcloud`'s snap whenever `nano` is updated for security patches. Fortunately, `nano`'s track record isn't that bad with 3 tracked CVEs: https://ubuntu.com/security/cves?package=nano

Nano makes the most sense due to space consideration.

simondeziel commented:
@roosterfish in https://chat.canonical.com/canonical/pl/bqbb3p4bc38tmn4ppe5y9k7q6o, it was suggested to try:
> I'd actually try to set `EDITOR: xdg-open` in your snapcraft.yaml in an `environment:` stanza (no idea if that works but this would be my first attempt)

roosterfish commented:
> I'd actually try to set `EDITOR: xdg-open` in your snapcraft.yaml in an `environment:` stanza

This returns almost the same: `Error: fork/exec /usr/bin/xdg-open: permission denied`.
Doesn't matter if defined in `apps.microcloud.environment` or as `export EDITOR` in the command right before exec'ing `microcloudd`.

simondeziel commented:
@roosterfish sorry to keep this saga alive but I just rediscovered that `vim-tiny` is shipped inside `coreXX` snaps (https://github.com/canonical/lxd-pkg-snap/pull/237):

```
$ ll /snap/core2?/current/usr/bin/vim.tiny 
-rwxr-xr-x 1 root root 1520320 Sep 25 01:30 /snap/core22/current/usr/bin/vim.tiny*
-rwxr-xr-x 1 root root 1736392 Aug 27 00:08 /snap/core24/current/usr/bin/vim.tiny*
```

Making use of it would essentially be free in terms of snap maintenance and space.

masnax commented:
> @roosterfish sorry to keep this saga alive but I just rediscovered that `vim-tiny` is shipped inside `coreXX` snaps ([canonical/lxd-pkg-snap#237](https://github.com/canonical/lxd-pkg-snap/pull/237)):
> 
> ```
> $ ll /snap/core2?/current/usr/bin/vim.tiny 
> -rwxr-xr-x 1 root root 1520320 Sep 25 01:30 /snap/core22/current/usr/bin/vim.tiny*
> -rwxr-xr-x 1 root root 1736392 Aug 27 00:08 /snap/core24/current/usr/bin/vim.tiny*
> ```
> 
> Making use of it would essentially be free in terms of snap maintenance and space.

As someone that hates using nano, I am in favour of this change :)

roosterfish commented:
> I just rediscovered that vim-tiny is shipped inside coreXX snaps

That sounds promising, but looks we don't have access to this from within the MicroCloud snap:

`Error: fork/exec /snap/core24/current/usr/bin/vim.tiny: permission denied`

> As someone that hates using nano, I am in favour of this change :)

I guess you have to be strong now!

simondeziel commented:
> > I just rediscovered that vim-tiny is shipped inside coreXX snaps
> 
> That sounds promising, but looks we don't have access to this from within the MicroCloud snap:
> 
> `Error: fork/exec /snap/core24/current/usr/bin/vim.tiny: permission denied`

I have a hard time reasoning about this... We outta be able to use binaries from the base snap otherwise MicroCloud's hooks/ scripts wouldn't have any `/bin/sh` or `ls` to work with. I'd be curious to take a look at the test env, if possible.

roosterfish commented:
> I have a hard time reasoning about this... 

I can step you through the reproducer:

1. Put `export EDITOR="/snap/core24/current/usr/bin/vim.tiny" in `commands/microcloud`.
2. `snapcraft`
3. `snap install <build-name> --dangerous`
4. Deploy a single node MicroCloud with `microcloud init`
5. `microcloud cluster recover` and answer with `yes`

Update:

Same error if you just do `EDITOR="vim.tiny"`, only the path in the error is then different.

simondeziel commented:
@roosterfish thanks for the reproducer. FYI, to test small changes like that, rebuilding/installing the snap is not needed, this is easier:

```
$ cp /snap/microcloud/current/commands/microcloud .
$ vim microcloud  # make the needed changes
$ sudo mount -o ro,bind microcloud /snap/microcloud/current/commands/microcloud
$ microcloud cluster recover
```

If you are editing a `daemon.start` file, know that you need to stop/start before after the mount bind.

So back to that `vim.tiny`. It turns out that I was on the right track with `ls` being similar as coming from the `coreXX` snap too. It's just that `ls` was part of the allow list which is missing `vim.tiny`... https://github.com/canonical/snapd/pull/14743

roosterfish commented:
> It's just that ls was part of the allow list which is missing vim.tiny

Nice one getting allowance for this in snapd directly! Seems to be way more robust going forward.
I have saved the code snippet, wasn't aware of this, appreciate it.

Regarding the next steps I think it will take some time until this is released (will it be even backported to older versions of snapd?) so do we want to continue on with our own bundled editor for now? Maybe switching to `vim.tiny` instead of `nano` will allow us to just swap the binary at a later point in time without modifying the UX.

simondeziel commented:
> Regarding the next steps I think it will take some time until this is released (will it be even backported to older versions of snapd?)

Yes, snapd is always backported to all supported releases, usually all at once.

> so do we want to continue on with our own bundled editor for now? Maybe switching to `vim.tiny` instead of `nano` will allow us to just swap the binary at a later point in time without modifying the UX.

Yes, let's add a `vim` part to our snap for the interim. Thanks

roosterfish committed:
snapcraft/commands/microcloud: Run the CLI from an accessible directory

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish commented:
@simondeziel I have now updated the PR to use `vim.tiny`.

But it looks `-Z --clean` aren't enough to prevent VIM trying to access/load something from your present working directory which causes `DENIED` apparmor messages.
To prevent this I have changed the `$PWD` of the CLI to `$SNAP_USER_DATA` which allows reads by `vim.tiny` when launched.

Furthermore there looks to be a [bug](https://bugs.launchpad.net/ubuntu/+source/vim/+bug/1968912) with `vim.tiny` which still tries to read a `defaults.vim` file regardless of the `--clean` setting. To prevent this I am priming an empty config file and moving `$VIMRUNTIME` to the corresponding snap path.
Otherwise you are always prompted with `E1187: Failed to source defaults.vim` when running `microcloud cluster recover` which doesn't feel right.

simondeziel commented on snapcraft.yaml:
```diff
@@ -84,10 +84,23 @@ parts:
+      override-build: |
+        touch "${CRAFT_PART_INSTALL}/usr/share/vim/vim91/defaults.vim"
+      organize:
+        usr/bin: bin/
```
Back when LXD was bundling `vim.tiny`, we also had this organize line:
```
usr/share/vim/vim*/debian.vim: etc/vimrc
```

Dunno if that would prevent to try and load the `defaults.vim` file that was never primed in the LXD case.

roosterfish replied:
I made some tests based on what I found in https://github.com/canonical/lxd-pkg-snap/pull/237. 
When you don't set `--clean`, then VIM tries to go through the entire chain of directories to find a config file. There is `-V` which gives you a nice rundown of probed files/dirs:

```
root@micro01:~# vim.tiny -V -Z /tmp/test
chdir(/tmp)
fchdir() to previous dir
chdir(/etc/vim)
fchdir() to previous dir
sourcing "/etc/vim/vimrc"
chdir(/usr/share/vim/vim91)
fchdir() to previous dir
line 23: sourcing "/usr/share/vim/vim91/debian.vim"
finished sourcing /usr/share/vim/vim91/debian.vim
continuing in /etc/vim/vimrc
finished sourcing /etc/vim/vimrc
chdir(/root)
fchdir() to previous dir
could not source "$HOME/.vimrc"
chdir(/root/.vim)
fchdir() to previous dir
could not source "~/.vim/vimrc"
chdir(/root)
fchdir() to previous dir
could not source "$HOME/.exrc"
chdir(/usr/share/vim/vim91)
fchdir() to previous dir
could not source "$VIMRUNTIME/defaults.vim"
Press ENTER or type command to continue
```

Note the last line `could not source "$VIMRUNTIME/defaults.vim"` from which I got the idea of just setting `$VIMRUNTIME` to a custom path.

So if we set `--clean` we can enforce not loading any configs but interestingly `defaults.vim` is still being probed:

```
root@micro01:~# vim.tiny -V -Z --clean /tmp/test
chdir(/tmp)
fchdir() to previous dir
chdir(/usr/share/vim/vim91)
fchdir() to previous dir
could not source "$VIMRUNTIME/defaults.vim"
E1187: Failed to source defaults.vim
Press ENTER or type command to continue
```

simondeziel replied:
> I made some tests based on what I found in [canonical/lxd-pkg-snap#237](https://github.com/canonical/lxd-pkg-snap/pull/237). When you don't set `--clean`, then VIM tries to go through the entire chain of directories to find a config file. 

Maybe that's was it was either using `VIMINIT` or `--clean`?

>There is `-V` which gives you a nice rundown of probed files/dirs:

Oh nice, could be handy!

roosterfish replied:
> Maybe that's was it was either using VIMINIT or --clean?

Even when setting `$VIMINT=""` it goes through the chain of locations to find any config file. So setting `--clean` brings us the farthest and it skips whatever you have set in `$VIMINT`.

simondeziel replied:
Have you tried `VIMINIT` pointing to the file you wanted it to source without looking for it?

roosterfish replied:
Yes when setting e.g. `VIMINIT="source /tmp/defaults.vim"` the error also doesn't appear but VIM is still looking into `/usr/share/vim/*` and `/etc/vim/*`:

```
root@micro02:~# VIMINIT="source /tmp/defaults.vim" vim.tiny -Z -V /tmp/test
chdir(/tmp)
fchdir() to previous dir
chdir(/etc/vim)
fchdir() to previous dir
sourcing "/etc/vim/vimrc"
chdir(/usr/share/vim/vim91)
fchdir() to previous dir
line 23: sourcing "/usr/share/vim/vim91/debian.vim"
finished sourcing /usr/share/vim/vim91/debian.vim
continuing in /etc/vim/vimrc
finished sourcing /etc/vim/vimrc
chdir(/tmp)
fchdir() to previous dir
line 0: sourcing "/tmp/defaults.vim"
finished sourcing /tmp/defaults.vim
continuing in VIMINIT
Press ENTER or type command to continue
```

And as soon we append `--clean` whatever is put for `VIMINIT` seems to be ignored and we are back at the error.

simondeziel replied:
Reading your `-V` output, it seems the `/etc/vim/vimrc` is the one providing a `runtimepath` to search through.
Anyway, thanks for making all those attempts/trials but I think we can just go with your currently working solution, I was mostly curious :)

roosterfish replied:
`/etc/vim/vimrc` doesn't exist in this example :) 
And IIRC `/usr/share/vim/vim91` is the default of `$VIMRUNTIME`.

Yeah I think we have a somewhat stable solution now and no apparmor DENIED messages.

simondeziel commented on snapcraft/commands/microcloud:
```diff
@@ -7,4 +7,14 @@ if [ -x "${SNAP_COMMON}/microcloud.debug" ]; then
+# which is caused by not having the 'vim-runtime' package.
+# Also see https://bugs.launchpad.net/ubuntu/+source/vim/+bug/1968912.
+export VIMRUNTIME="/snap/microcloud/current/usr/share/vim/vim91"
+export EDITOR="vim.tiny -Z --clean"
```
In the LXD case, `snapcraft/wrappers/editor` was also dealing with `VIMINIT` to point `vim.tiny` to the `/etc/vimrc` file.

roosterfish replied:
I tried, but same outcome as in https://github.com/canonical/microcloud-pkg-snap/pull/33#discussion_r1852199042, the `defaults.vim` is still probed.

simondeziel reviewed (approved):
I provided some historical data from the previous experience with LXD but if the `VIMRUNTIME` trick you found does the job, there is no point in changing it... especially for a part that won't be kept permanently.

Thanks for adding a pointer to the LP bug!
